lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -242,54 +195,111 @@ public void OnNext(IActiveContext activeContext)
         /// <param name="completedTask">The link to the completed task</param>
         public void OnNext(ICompletedTask completedTask)
         {
-            _completedTasks.Add(completedTask);
-
-            if (_completedTasks.Count != _dataSet.Count + 1)
-            {
-                return;
-            }
-            
-            foreach (var task in _completedTasks)
+            lock (_lock)
             {
-                Logger.Log(Level.Verbose, string.Format("Disposing task: {0}", task.Id));
-                task.ActiveContext.Dispose();
+                Logger.Log(Level.Info,
+                    string.Format("Received completed task message from task Id: {0}", completedTask.Id));
+                _completedTasks.Add(completedTask);
+
+                if (_completedTasks.Count != _totalMappers + 1)
+                {
+                    return;
+                }
+
+                _imruDone = true;
+                foreach (var task in _completedTasks)
+                {
+                    Logger.Log(Level.Info, string.Format("Disposing task: {0}", task.Id));
+                    task.ActiveContext.Dispose();
+                }
             }
         }
 
+        /// <summary>
+        /// Specifies what to do when evaluator fails.
+        /// If we get all completed tasks then ignore the failure
+        /// Else request a new evaluator. If failure happens in middle of IMRU 
+        /// job we expect neighboring evaluators to fail while doing 
+        /// communication and will use FailedTask and FailedContext logic to 
+        /// order shutdown.
+        /// </summary>
+        /// <param name="value"></param>
         public void OnNext(IFailedEvaluator value)
         {
-            Logger.Log(Level.Info, "An evaluator failed, checking if it failed before context and service was submitted");
-            int currFailedEvaluators = Interlocked.Increment(ref _currentFailedEvaluators);
-
-            if (value.FailedContexts != null && value.FailedContexts.Count != 0)
+            if (_imruDone)
```

Comment:
You might need a lock around `_imruDone`. Also, maybe find a better name for it?